polygeist/lib/polygeist/Passes/ParallelLICM.cpp

```diff
@@ -28,305 +31,365 @@ struct ParallelLICM : public ParallelLICMBase<ParallelLICM> {
-          if (isa<MemoryEffects::Write>(effect.getEffect()))
+        cast<MemoryEffectOpInterface>(other).getEffectsOnResource(res, effects);
+        for (const MemoryEffects::EffectInstance &effect : effects) {
+          if (isa<MemoryEffects::Allocate>(effect.getEffect()) ||
+              isa<MemoryEffects::Read>(effect.getEffect()) ||
+              isa<MemoryEffects::Write>(effect.getEffect()))
             return true;
         }
       }
-      for (auto res : writeResources) {
+
+      // Check whether the given operation read, writes or allocates a resources
+      // that is freed by the operation associated with this class.
+      for (SideEffects::Resource *res : freeResources) {
         SmallVector<MemoryEffects::EffectInstance> effects;
-        memEffect.getEffectsOnResource(res, effects);
-        for (auto effect : effects) {
-          if (isa<MemoryEffects::Allocate>(effect.getEffect()))
-            return true;
-          if (isa<MemoryEffects::Read>(effect.getEffect()))
+        cast<MemoryEffectOpInterface>(other).getEffectsOnResource(res, effects);
+        for (const MemoryEffects::EffectInstance &effect : effects) {
+          if (isa<MemoryEffects::Allocate>(effect.getEffect()) ||
+              isa<MemoryEffects::Write>(effect.getEffect()) ||
+              isa<MemoryEffects::Read>(effect.getEffect()))
             return true;
         }
       }
-      for (auto res : freeResources) {
-        SmallVector<MemoryEffects::EffectInstance> effects;
-        memEffect.getEffectsOnResource(res, effects);
-        for (auto effect : effects) {
-          if (isa<MemoryEffects::Allocate>(effect.getEffect()))
-            return true;
-          if (isa<MemoryEffects::Write>(effect.getEffect()))
-            return true;
-          if (isa<MemoryEffects::Read>(effect.getEffect()))
+    }
+
+    return false;
+  }
+
+  const SmallVector<SideEffects::Resource *> &getReadResources() const {
+    return readResources;
+  };
+  const SmallVector<SideEffects::Resource *> &getWriteResources() const {
+    return writeResources;
+  };
+  const SmallVector<SideEffects::Resource *> &getFreeResources() const {
+    return freeResources;
+  };
+  const SmallVector<SideEffects::Resource *> &getAllocateResources() const {
+    return allocateResources;
+  };
+
+private:
+  SmallVector<SideEffects::Resource *> readResources;
+  SmallVector<SideEffects::Resource *> writeResources;
+  SmallVector<SideEffects::Resource *> freeResources;
+  SmallVector<SideEffects::Resource *> allocateResources;
+};
+
+/// Returns true if the Operation \p op can be hoisted out of the given loop \p
+/// loop. The \p willBeMoved argument represents operations that are known to be
+/// loop invariant (and therefore will be moved outside of the loop).
+template <typename T, typename = std::enable_if_t<llvm::is_one_of<
+                          T, scf::ParallelOp, AffineParallelOp>::value>>
+static bool canBeHoisted(Operation &op, T loop,
```

Comment:
Probably good to keep the original name (`canBeParallelHoisted`), as the code has assumption on the loop being parallel loop, some checks are not safe if the loop is not a parallel loop.

Reply:
It is clear that this function can only support parallel loops, the constraint on the typename enforces that condition and the new name is shorter.